evm/p2p/kademlia.py

```diff
@@ -264,6 +268,22 @@ def neighbours(self, node_id, k=k_bucket_size):
         return sort_by_distance(nodes, node_id)[:k]
 
 
+def binary_get_bucket_for_node(buckets, node):
+    """Return the bucket for a given node."""
+    sorted_buckets = sorted(buckets)
+    bucket_ends = [bucket.end for bucket in sorted_buckets]
+    bucket_position = bisect.bisect_left(bucket_ends, node.id)
+    # Prevents edge cases where bisect_left returns an out of range index
+    try:
+        bucket = sorted_buckets[bucket_position]
+    except IndexError:
+        raise ValueError("No bucket found for node with id {}".format(node.id))
+    bucket = sorted_buckets[bucket_position]
+    if not bucket.start <= node.id <= bucket.end:
+        raise ValueError("No bucket found for node with id {}".format(node.id))
```

Comment:
How can this happen?

Reply:
Well, I was thinking about if a list of buckets was input that had breaks - i.e. `KBucket(0,2)` and `KBucket(5,10)` and node.id was 4. But it looks as though that's not possible with how split_bucket is implemented?

Reply:
Yeah, that's not possible given how KBucket.split() is implemented, but more importantly, if the position returned by bisect exists on the list, this check would never fail, would it?

Reply:
The only reason I can think of is if `node.id` is for some reason below the lowest `bucket.start` or above the highest `bucket.end`, though I'm not sure how likely that is. If it's not a possible edge case, I can remove this check.